src/zmalloc.c

```diff
@@ -346,42 +347,50 @@ static inline void *ztryrealloc_usable_internal(void *ptr, size_t size, size_t *
 
     /* not allocating anything, just redirect to free. */
     if (size == 0 && ptr != NULL) {
-        zfree(ptr);
+        zfree_usable(ptr, &oldsize);
         if (usable) *usable = 0;
+        if (old_usable) *old_usable = oldsize;
         return NULL;
     }
     /* Not freeing anything, just redirect to malloc. */
-    if (ptr == NULL)
+    if (ptr == NULL) {
+        if (old_usable) *old_usable = 0;
         return ztrymalloc_usable(size, usable);
+    }
 
     /* Possible overflow, return NULL, so that the caller can panic or handle a failed allocation. */
     if (size >= SIZE_MAX/2) {
-        zfree(ptr);
+        zfree_usable(ptr, &oldsize);
         if (usable) *usable = 0;
+        if (old_usable) *usable = oldsize;
```

Comment:
```suggestion
        if (old_usable) *old_usable = oldsize;
```

.... there are many such conditions. maybe instead do it once at start:
```
    size_t dummy;    
    if (old_usable == NULL) old_usable = &dummy;
    if (usable == NULL) usable = &dummy;
```

Reply:
Done in [cbbac14](https://github.com/redis/redis/pull/14363/commits/cbbac14721320252f0bc3bc25b1e707b65b0f6e0)